### PR TITLE
Limit FeeRate change for the UpdateFee msg to prevent sharp changes

### PR DESCRIPTION
**Pull request by ziggie1984**
Relates to #7666 and #7756

## Detailed Problem explanation:

In lnd (lightning) there exist the feeupdate msg which allows channel openers to change the fee of the commitment transaction. For STATIC_REMOTE_KEY channels HTLCs which are part of the Commitment Transaction do have Timeout/Success Transactions which are signed with the same feerate as the Commitment Transaction (see BOLT3 for more details about the HTLC Transaction). 
The FeeUpdate is only allowed to be sent by the Channel Opener but it should be in acceptable Limits, because especially for non-achor channels there is no way to accelerate this transaction (only if you pay a miner directly). CLN nodes will check for a specific range and will not allow the feeupdate from happening. If CLN does reject the feeupdate the channel will either be force-closed by the peer (lndk) or the channel becomes unusable and needs to be force-closed eventually by the peer.

This PR fixes 2 Problems:

### No valid Fee Estimation by our backend

In case you either wipe your mempool or restart a new bitcoind node you want to connect to your lnd node, there is a short time period where bitcoind is not reporting any fee estimations because it does not have enough data available.

It will report: 
```
bitcoin-cli estimatesmartfee 3
{
  "errors": [
    "Insufficient data or no feerate found"
  ],
  "blocks": 0
}
```
In such cases lnd would just read this response and the fee estimation would remain zero and then would just enforce the min mempool fee instead. This is a bad choice because when our mempool is big enough this fee rate is always 1 sat/vbyte and does not represent the current fee rate conditions. This PR will return an error in such cases and lnd will use its fallback feerate of 25 sats/vbyte. It is always better to overestimate rather than underestimate in such cases. 

### Limit Fee Estimation when channel is drained locally

Moreover Lnd reevaluates the MaxFeeRate it is willing to allocate for a Channel every time a feeupdate  msg will be sent. When the channel is locally drained this could lead to a situation where we decrease the feerate all the way to the fee floor which does not reflect the current bitcoin network fee rate correctly and could also lead to our peer rejecting these low fee rate updates. With this PR we do not lower the MaxFeeRate we are willing to allocate for this channel if it would fall below the old fee rate. This guarantees that if we had a high fee rate on a channel, that we do not decrease it when the local balance is drained.   

Locking for a Concept Ack here before writing tests.

This change will not fix existing channels which ran into this issue, but will most likely prevent future channels form being unusable when proposing very low fee rate updates leading to the channel-peer rejecting those.

## Change Description
~~This PR limits the FeeRate update to a maximum of 30% of the old feerate. It only applies to non-anchor channels because for those the negotiated fee cannot be changed when resolving on-chain.~~

~~Open for discussions whether we want to pursue this route if 30% is too conservative and so one.~~

~~Needs some style fixing will solve this in the coming days.~~

## Tests 

I currently just updated the unit tests in the `lnwallet` package. I think it is enough and testing this in the `htlcswitch` where we drain the channel locally and send consecutively fee updates with the default fee-allocation might not be worth it IMO, what are your takes?

Fixes https://github.com/lightningnetwork/lnd/issues/7756

## Summary by CodeRabbit

- **Bug Fixes**
	- Fixed an issue where low fee rates could be proposed for channel transactions, ensuring alignment with network standards.
	- Adjusted fee buffer handling in channel openings to comply with BOLT 02 specification.
	- Improved fee estimation when external data is insufficient.
- **New Features**
	- Added support for Taproot witness types in RPC calls.
- **Documentation**
	- Updated handling instructions for `min_final_cltv_expiry_delta` for external invoices.
- **Tests**
	- Enhanced testing for channel fee rate calculations to cover more scenarios.

### CODE REVIEW

**Comment by ziggie1984** [comment]
This concept will not fix the problem, stated in https://github.com/lightningnetwork/lnd/issues/7756#issuecomment-1638969967. I need to rework this to make sure we do not scale down the comitment fee all the way to 253 sat/kw. 

Hold on for review yet !

File: lnwallet/chainfee/estimator.go
Comment (kornpow) [inline lnwallet/chainfee/estimator.go:482]: why update these log messages?

**Comment by kornpow** [comment]
Will this PR fix stuck channels or will it only prevent channels from entering a stuck state in the future?

File: lnwallet/chainfee/estimator.go
Comment (ziggie1984) [reply lnwallet/chainfee/estimator.go:482]: because the unit in itself `SatPerKWeight` has a string method which already prints out the sat/kw

**Comment by ziggie1984** [comment]
> Will this PR fix stuck channels or will it only prevent channels from entering a stuck state in the future?

It cannot undo the feeupdate which happend in the past (your case currently) tho it will prevent low feeupdates in the future.

> This change will not fix existing channels which ran into this issue, but will most likely prevent future channels form being unusable when proposing very low fee rate updates leading to the channel-peer rejecting those.

File: lnwallet/channel.go
Comment (yyforyongyu) [inline lnwallet/channel.go:7354]: A bit confused here - so this means our fee rate can only go up not down? And if we don't have this local balance, how can we cover the fees here?

Comment (ziggie1984) [reply lnwallet/channel.go:7354]: exactly we do not want to lower the max_feerate lower than the old_fee, because this could lead to a situation where we go all the way to 253 sats/kw, we want to decrease the feerate here: https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L8456-L8458, basically we want to either have a 3 block sample_feerate which is lower than the old_feerate before we allow a lower feeupdate, otherwise it makes not real sense to lower the MaxFeeRate.

**Comment by ziggie1984** [comment]
Suffered from this behaviour on my local node as well now, I think this should be addressed asap, Channel is unusable now until we reach lower feelimits.

2023-12-10 13:30:59.882 [WRN] HSWC: ChannelLink(aa4a1c1313bed1063a9529840062c3263aced0b544b28105e79019f91e9c2bba:2): received warning message from peer: chan_id=ba2b9c1ef91990e70581b244b5d0ce3a26c362008429953a06d1be13131c4aa8, err=update_fee 253 outside range 4724-702240 (currently 1795)

File: lnwallet/channel.go
Comment (yyforyongyu) [reply lnwallet/channel.go:8457]: In that case I think we should compare it against 3 block sample feerate instead of old feerate, otherwise if the fee spikes for a while then goes down, we'd still use the higher feerate as the max feerate although the current mempool is less congested. Plus this would bypass the value `maxAllocation` configured by the users?

**Review by yyforyongyu — COMMENTED** [review]
I think if we have #8096 we don't need this anymore since we would then have a fee buffer?

File: lnwallet/channel.go
Comment (ziggie1984) [reply lnwallet/channel.go:8457]: Hmm not sure if I understand the detail of your proposal could you elaborate ?

I think binding the maxAllocation to our current amount we have might cause this strange logic problem.

What about refering the `maxAllocation` do the ChannelCapacity, I think that makes more sense because as a channelopener you will have most of the balance on your side anyways at the beginning and are willing to pay this max amount anyways. Now when the channel becomes depeleted the maxFeeRate is always (`localbalance+oldfee`) but if we have a lot of balance on our side it would be capped at the (`chancapacity *MaxAllocation`) ?

Comment (ziggie1984) [reply lnwallet/channel.go:8457]: We would be protected against big feespikes, not allocating more that our initial willingess to pay, but if the channel is depleted we make sure we do not decrease our MaxFee the more our channel gets depleted hence reducing the local fee of the commitment all the time, that also not fair to the peer imo which now has all the balance but a crappy feerate.

**Review by yyforyongyu — COMMENTED** [review]
Looking good, just a few nits. One thing I think we should do in a followup PR is to make it more clear on how we update fees for anchor vs non-anchor channels here:
https://github.com/lightningnetwork/lnd/blob/4d8fa349ca405a42548717c17cc689e6612ff37a/htlcswitch/link.go#L1140

Since for anchor channels we only need to make sure we pay enough fees to relay.

**Comment by ziggie1984** [comment]
> Looking good, just a few nits. One thing I think we should do in a followup PR is to make it more clear on how we update fees for anchor vs non-anchor channels here:

Agree I think this relates to https://github.com/lightningnetwork/lnd/issues/8302, which should get rid of the `MaxAnchorsCommitFeeRate` config setting and always make sure the relay feerate of its average Peers are met.

File: lnwallet/chainfee/estimator.go
Comment (ziggie1984) [reply lnwallet/chainfee/estimator.go:482]: Sounds reasonable, I wonder whether we need to Refactor `bitcoindConn *rpcclient.Client` to interface it out so that we can mock the rpc call, or do you maybe have other idea ?

File: lnwallet/channel.go
Comment (ziggie1984) [reply lnwallet/channel.go:8459]: Yes I think we can go with the current approach, but I think the FeeRate for the Commitment Tx is really not an easy to solve problem. So I think making sure that the Commitment has always the Relay Fee of its peer is a very important PR which should be prioritized.

File: lnwallet/chainfee/estimator.go
Comment (yyforyongyu) [reply lnwallet/chainfee/estimator.go:482]: It'd be great to have an interface there, yes, but maybe a future TODO to limit the scope.

File: lnwallet/channel.go
Comment (yyforyongyu) [inline lnwallet/channel.go:8465]: I don't think we should do this check anymore as we are using at least `oldFeeRate` and, if that value is below fee floor, then it means there's a bug in other subsystems, such as fee estimator - in that case we should fix it there. Defaulting to fee floor here will hide that bug. Also means we can remove `MaxSatPerKweight` for now?

File: lnwallet/channel_test.go
Comment (yyforyongyu) [inline lnwallet/channel_test.go:7532]: how did you get this `6000` as the old feerate?

File: lnwallet/channel_test.go
Comment (yyforyongyu) [inline lnwallet/channel_test.go:7550]: again would be nice to have some docs here explaining the value `5999`.

**Review by yyforyongyu — COMMENTED** [review]
Looking good! Just a few nits and questions.

File: lnwallet/channel.go
Comment (ziggie1984) [reply lnwallet/channel.go]: renamed to `currentFee` instead of `oldFee` and also removed the check.

File: lnwallet/chainfee/rates.go
Comment (yyforyongyu) [inline lnwallet/chainfee/rates.go:84]: This is not used anywhere tho

**Review by yyforyongyu — COMMENTED** [review]
Cool, just an unused function, and needs a rebase!

**Review by yyforyongyu — APPROVED** [review]
LGTM🏖

**Comment by kornpow** [comment]
🚢 would be great for this ship to set sail

**Comment by yyforyongyu** [comment]
need a second reviewer 👀  cc @saubyk

File: lnwallet/channel.go
Comment (ProofOfKeags) [inline lnwallet/channel.go:8704]: can we avoid mutation here either by inlining the original maxFee definition or renaming the original maxFee to something like maxFeeProposal?

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7565]: The magic numbers in these tests make me uncomfortable. I had to drill into the guts of the test setup to convince myself that 6000 was the right number.

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7592]: Can we restructure this test by getting it to reveal it's current fee allocation and comparing against that? This number is even more magic than the last one and while the comments indicate to me that this number, magic as it is, is intentional, it doesn't actually do much to persuade me that the property being checked (fees always ratchet upwards) is actually holding.

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7633]: Correct me if I'm wrong, but doesn't this just duplicate the logic implemented inside the LightningChannel object?

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7563]: s/definied/defined/g

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7586]: s/definied/defined/g

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7680]: This test structure doesn't give me a lot of confidence overall. I understand that it precedes your changes here, but I think it's too fragile. Not sure it's worth fixing here though. What I'd like to see instead is the emergent hierarchy of fee choices as one test, and the ratcheting property you implement here as a separate test altogether.

**Review by ProofOfKeags — CHANGES_REQUESTED** [review]
Overall the change itself seems fine and the approach seems alright. I'm slightly skeptical about it as it can result in cases where the user will overpay for fees but I'm willing to accept that tradeoff and see how this fares.

My criticism centers on the testing strategy. It seems that a lot of the testing strategy precedes your changes so I feel reluctant to make redoing this corner of the test suite a requirement for approval, but I would at least like to see your tests separated out more and have the core property of fee ratcheting come to the forefront of the test structure.

File: lnwallet/channel_test.go
Comment (ziggie1984) [reply lnwallet/channel_test.go]: yes you are right I changed it, now I use this function to calculate the max_feerate without expecting the ratcheting property

File: lnwallet/channel_test.go
Comment (ziggie1984) [reply lnwallet/channel_test.go]: I tried to reorganize the tests, without anticipating the ratcheting property. I think it is better to keep all the test possibilities in one place but PTAL.

**Comment by ziggie1984** [comment]
@coderabbitai review

File: lnwallet/channel_test.go
Comment (ziggie1984) [reply lnwallet/channel_test.go]: I am now comparing the fee allocation values to avoid the magic numbers wdyt ?

File: lnwallet/channel.go
Comment (ProofOfKeags) [inline lnwallet/channel.go:8684]: Please add to the godoc of this stating that the second return value is a fraction value [0,1] representing the share of the total balance that is allocated to fees. I would not ordinarily be able to ascertain this without inspecting the definition.

File: lnwallet/channel.go
Comment (ProofOfKeags) [inline lnwallet/channel.go:8712]: I'd prefer to take a more functional approach using [Max](https://pkg.go.dev/math#Max). From there it follows that the feeAllocation is always `maxFee / baseBalance`. This avoids having to track the changing semantics through this if block.

```go
idealMaxFee := float64(baseBalance) * maxAllocation
maxFee := math.Max(float64(currentFee), idealMaxFee)
maxFeeAllocation := maxFee / float64(baseBalance)
maxFeeRate := ...
return maxFeeRate, maxFeeAllocation
```

With this style, each name has a *singular semantic interpretation*.

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7549]: use `require.NoError`. We're trying to phase out the explicit error blocks.

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7636]: 🫡

Comment (ProofOfKeags) [reply lnwallet/channel_test.go:7636]: I'm curious how far we can crank down that 0.01 before the tests start failing. I suspect we can get down to 1e-8

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7710]: `require.NoError`

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7809]: 🫡

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [inline lnwallet/channel_test.go:7561]: I think this is a decent holdover for the time being. The numbers are still magic but restructuring these tests is not the priority for the moment so I think this is a great intermediate result. 🤝

**Review by ProofOfKeags — CHANGES_REQUESTED** [review]
I'm happy with this PR with the exception of some nits. I think if we just clean up the rough edges, I think we're ready to ship this. Thank you for the submission! 🧡

File: lnwallet/channel_test.go
Comment (ziggie1984) [reply lnwallet/channel_test.go:7636]: 0.1% is the lowest we can go, otherwise 1/6000 => 1.6e-4, so I changed it to 0.1%.

**Review by ProofOfKeags — APPROVED** [review]
SNED 🚀

**Comment by ziggie1984** [comment]
Last Push was just a rebase.

**Comment by ProofOfKeags** [comment]
Idk what's going on with CI but there's a lot of complaining checks that I can't figure out why they are failing.

**Comment by guggero** [comment]
Related to the build cache. Going to put up a fix to reduce the cache size.
@Crypt-iQ you self-requested review on this PR after 2 approvals. Should we block on your review?

**Comment by Crypt-iQ** [comment]
> Related to the build cache. Going to put up a fix to reduce the cache size. @Crypt-iQ you self-requested review on this PR after 2 approvals. Should we block on your review?

Yes

**Comment by guggero** [comment]
Needs a rebase.

**Pull request merged**
